SI: Vue3 module template

alejoroman0605 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This update will create a new VueJS template in Upendo DNN Generator that works almost identically to how existing VueJS and Angular templates are offered today.

## How Has This Been Tested?

- I created a clean DNN site.
- I created a module using the Vue3 template.
- I created a new page.
- I installed the module on the DNN site.
- I added the module to the page.

## Screenshots:
![image](https://github.com/UpendoVentures/generator-upendodnn/assets/48692645/46f74119-d2b0-446b-b62a-bbbdd1b8a869)

![image](https://github.com/UpendoVentures/generator-upendodnn/assets/48692645/25083481-85d6-4086-bb4b-089ac02d3a52)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

WillStrohl commented on generators/spa/templates/Vue3JS/common/Data/Item.cs:
```diff
@@ -0,0 +1,61 @@
+    /// </summary>
+    [PrimaryKey("Id", AutoIncrement = true)]
+    [Scope("ModuleId")]
+    public partial class Item
```
@alejoroman0605  It's great that there is a DAL example in here using an object, but it should be a generic example, like we have in the other module templates.

WillStrohl replied:
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I did it as it is in other Item.cs examples from other templates used.
https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/0980c63dc4abfd2a0bcc3cf82662d1a0a8126a0f

WillStrohl commented on generators/spa/templates/Vue3JS/common/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,34 @@
+//
+// You can specify all the values or you can default the Revision and Build Numbers 
+// by using the '*' as shown below:
+[assembly: AssemblyVersion("01.00.00")]
```
@alejoroman0605 There is a token present in the generator that should be used here.

WillStrohl replied:
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola  
https://github.com/UpendoVentures/generator-upendodnn/blob/master/generators/spa/index.js#L121

alejoroman0605 replied:
@WillStrohl I already updated https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/10c1a9a7c124c3e76beb87fba71ea268d3a06c63,
This is also wrong in the Vue 2 template, do I fix it? 
https://github.com/UpendoVentures/generator-upendodnn/blob/master/generators/spa/templates/VueJS/common/Properties/AssemblyInfo.cs#L33

WillStrohl replied:
@alejoroman0605 No need to update the other one too.  Instead, an issue should be created to fix that later.  That already existed before this update.  Also, I've notified you and the rest of the team about the new tokens after that, but before this.  You should have been aware already.

WillStrohl commented on generators/spa/templates/Vue3JS/common/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,34 @@
+// You can specify all the values or you can default the Revision and Build Numbers 
+// by using the '*' as shown below:
+[assembly: AssemblyVersion("01.00.00")]
+[assembly: AssemblyFileVersion("01.00.00")]
```
@alejoroman0605 There is a token present in the generator that should be used here.

WillStrohl replied:
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola  
https://github.com/UpendoVentures/generator-upendodnn/blob/master/generators/spa/index.js#L121

alejoroman0605 replied:
@WillStrohl I already updated 10c1a9a,
This is also wrong in the Vue 2 template, do I fix it?
https://github.com/UpendoVentures/generator-upendodnn/blob/master/generators/spa/templates/VueJS/common/Properties/AssemblyInfo.cs#L34

WillStrohl replied:
@alejoroman0605 Not yet. Please create a new issue for this.  That instance was there before this update and after the new tokens were introduced.

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Release/UpendoVentures.UpendoSendUI.csproj.CoreCompileInputs.cache:
```diff
@@ -0,0 +1 @@
+0bc7793b7e22ae5cc80ca1f786f579c15002d7a3
```
@alejoroman0605 This file should not be committed.

WillStrohl replied:
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/b667ca74420311f604a3735b866d4c842e9b08e0

WillStrohl commented on generators/spa/templates/Vue3JS/common/packages/MSBuildTasks.1.4.0.88/tools/Build.proj:
```diff
@@ -0,0 +1,65 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
@alejoroman0605 Why isn't this using the same build files as the other module templates?

WillStrohl commented on generators/spa/templates/Vue3JS/common/packages/MSBuildTasks.1.4.0.88/tools/MSBuild.Community.Tasks.Targets:
```diff
@@ -0,0 +1,155 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
@alejoroman0605 Why isn't this using the same build files as the other module templates?

WillStrohl commented on generators/spa/templates/Vue3JS/common/packages/MSBuildTasks.1.4.0.88/tools/MSBuild.psm1:
```diff
@@ -0,0 +1,146 @@
+function Resolve-ProjectName {
```
@alejoroman0605 Why isn't this using the same build files as the other module templates?

WillStrohl commented on generators/spa/templates/Vue3JS/common/packages/Microsoft.AspNet.WebApi.Core.5.2.3/Content/web.config.transform:
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
@alejoroman0605 Why isn't this using the same build files as the other module templates?

WillStrohl commented on generators/spa/templates/Vue3JS/common/packages/Microsoft.AspNet.WebApi.WebHost.5.2.3/lib/net45/System.Web.Http.WebHost.xml:
```diff
@@ -0,0 +1,135 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
@alejoroman0605 Why isn't this using the same build files as the other module templates?

WillStrohl commented on generators/spa/templates/Vue3JS/common/packages/Newtonsoft.Json.13.0.3/LICENSE.md:
```diff
@@ -0,0 +1,20 @@
+The MIT License (MIT)
```
@alejoroman0605 Why is this file committed?

WillStrohl commented on generators/spa/templates/Vue3JS/common/packages/Newtonsoft.Json.13.0.3/README.md:
```diff
@@ -0,0 +1,71 @@
+# ![Logo](https://raw.githubusercontent.com/JamesNK/Newtonsoft.Json/master/Doc/icons/logo.jpg) Json.NET
```
@alejoroman0605 Why is this file committed?

WillStrohl reviewed (changes requested):
@alejoroman0605 There are a lot of files committed that shouldn't be here, and the module is not following the correct development pattern.  It should function the same as the other modules.

alejoroman0605 commented:
@WillStrohl I have corrected the template addressing all the issues that were previously detected. To achieve this, I updated the .csproj file and included the Module.build so that it uses the same build files as the other module templates. Instead of relying on references to local files, I migrated to using NuGet to manage dependencies. Additionally, I completely removed the local files and their respective references from the project.

WillStrohl commented on generators/spa/index.js:
```diff
@@ -14,7 +14,8 @@ module.exports = class extends DnnGeneratorBase {
           { name: 'Angular', value: 'Angular' },
           { name: 'ReactJS', value: 'ReactJS' },
-          { name: 'VueJS', value: 'VueJS' }
+          { name: 'Vue 2', value: 'VueJS' },
```
@alejoroman0605 You ignored my note on this. I asked you to change this to `Vue 2` (if it's Vue 2.xx).  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I did not ignore this comment, what I did was change the name, which is what is displayed in the console, but I immediately changed the value as well.

WillStrohl replied:
@alejoroman0605 When I looked at this before your response, the values did not appear to be updated.

WillStrohl commented on generators/spa/index.js:
```diff
@@ -14,7 +14,8 @@ module.exports = class extends DnnGeneratorBase {
           { name: 'ReactJS', value: 'ReactJS' },
-          { name: 'VueJS', value: 'VueJS' }
+          { name: 'Vue 2', value: 'VueJS' },
+          { name: 'Vue 3', value: 'Vue3JS' }
```
@alejoroman0605 You ignored my note on this. I asked you to change this to `Vue 3`.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I did not ignore this comment, what I did was change the name, which is what is displayed in the console, but I immediately changed the value as well.

alejoroman0605 replied:
@WillStrohl The index.js uses the value to search the template folder, I modified this line: 
let spaPath = spaType === "ReactJS" ? `${this.props.spaType}/${this.props.langType}` : spaType === "Vue 2" ? "VueJS/" : spaType === "Vue 3" ? "Vue3JS/" : `${this.props.spaType}/`;
so it would find the template files, is that okay? Otherwise I would have to rename the VueJS and Vue3JS folder.

WillStrohl replied:
@alejoroman0605 When I looked at this before your response, the values did not appear to be updated.

WillStrohl commented on generators/spa/templates/Vue3JS/common/Components/BaseClasses/ApiControllerBase.cs:
```diff
@@ -0,0 +1,51 @@
+        /// </summary>
+        /// <typeparam name="T">The type of JSON token to return (e.g., JToken, JArray, or JObject).</typeparam>
+        /// <returns>The parsed JSON content as the specified type.</returns>
+        public T ContentJ<T>() where T : JToken
```
@alejoroman0605 You ignored this comment. CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl Not really, I'll delete it

WillStrohl commented on generators/spa/templates/Vue3JS/common/Components/Common.cs:
```diff
@@ -0,0 +1,48 @@
+    /// <summary>
+    /// A collection of common utility methods.
+    /// </summary>
+    public static class Common
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl Not really, I'll delete it

WillStrohl commented on generators/spa/templates/Vue3JS/common/Components/FeatureController.cs:
```diff
@@ -0,0 +1,20 @@
+    /// <summary>
+    /// The controller class responsible for managing module features and upgrades.
+    /// </summary>
+    public sealed class FeatureController : IUpgradeable
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I already used the pattern found in the module templates above.

WillStrohl commented on generators/spa/templates/Vue3JS/common/Components/FeatureController.cs:
```diff
@@ -0,0 +1,20 @@
+        /// </summary>
+        /// <param name="Version">The version to which the module is being upgraded.</param>
+        /// <returns>A message indicating the result of the upgrade process.</returns>
+        public string UpgradeModule(string Version)
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I already used the pattern found in the module templates above.

WillStrohl commented on generators/spa/templates/Vue3JS/common/Constants/ModuleConstants.cs:
```diff
@@ -0,0 +1,48 @@
+    /// <summary>
+    /// Constants used within the <%= friendlyName %> module.
+    /// </summary>
+    public class ModuleConstants
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I uploaded this class because I use constants in the code.
example:
```
  [HttpGet]
        [ActionName("GetResx")]
        public HttpResponseMessage GetResx(string filename)
        {
            var resx = new JObject();

            var resxRoot = $"{ModuleConstants.ResxPartialRoot}{filename}.resx";
            var filepath = HttpContext.Current.Server.MapPath(resxRoot);
            var resxDoc = new XmlDocument();
            resxDoc.Load(filepath);

            foreach (XmlNode dataNode in resxDoc.DocumentElement.SelectNodes(ModuleConstants.RootData))
            {
                var key = dataNode.Attributes[ModuleConstants.DataNodeAttributesName].Value;
                var val = Localization.GetString(key.ToString(), resxRoot);

                if (key.EndsWith(ModuleConstants.PointText, StringComparison.InvariantCultureIgnoreCase)) key = key.Substring(0, key.Length - 5);
                key = key.Replace(".", "_");

                resx.Add(key, val);
            }

            return Request.CreateResponse(resx);
        }
```

alejoroman0605 replied:
@WillStrohl I use that class in the code for default values such as:
 ```
public HttpResponseMessage GetResx(string filename)
        {
            var resx = new JObject();

            var resxRoot = $"{ModuleConstants.ResxPartialRoot}{filename}.resx";
            var filepath = HttpContext.Current.Server.MapPath(resxRoot);
            var resxDoc = new XmlDocument();
            resxDoc.Load(filepath);

            foreach (XmlNode dataNode in resxDoc.DocumentElement.SelectNodes(ModuleConstants.RootData))
```

WillStrohl replied:
Okay. As long as it's used. 👍🏽

WillStrohl commented on generators/spa/templates/Vue3JS/common/Services/ViewModels/ModuleConfigModel.cs:
```diff
@@ -0,0 +1,12 @@
+    /// <summary>
+    /// Represents the configuration model for the <%= friendlyName %> module.
+    /// </summary>
+    public class ModuleConfigModel
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl Not really, I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/087a698623f5229c862b4fcf40a79c8fd96f33a0

WillStrohl commented on generators/spa/templates/Vue3JS/common/client-app/.gitignore:
```diff
@@ -0,0 +1,23 @@
+.DS_Store
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola  (This should be handled by the `.gitignore` that's included in the `Solution Structure` template, in the root of the solution.)

alejoroman0605 replied:
@WillStrohl I just fixed it. https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/b18dcbe25be458f59275176bc8102d38859aa4ae

WillStrohl commented on generators/spa/templates/Vue3JS/common/client-app/package.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "client-app",
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I updated to the latest versions of each of the dependencies. https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/bd5943156de65218efa701bba96fc272f7973057 , https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/6d998f503d0ff12e6291e18fbee5515bc8c7258c
![image](https://github.com/UpendoVentures/generator-upendodnn/assets/48692645/a1119904-fb09-4b9f-826d-09eaea0d7f1a)

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Debug/Module.csproj.FileListAbsolute.txt:
```diff
@@ -0,0 +1,7 @@
+D:\Trabajo\NTSPRINT\Extension Generator Upendo DNN\Modules\Vue3JS\bin\Newtonsoft.Json.dll
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/cd7ec4884a2068511e4a9f0ebe887b68fccf1b26

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Debug/UpendoVentures.ModuleVue3.csproj.FileListAbsolute.txt:
```diff
@@ -0,0 +1,18 @@
+C:\Users\Nariel\Documents\Upendo\Modules\Upendo\UpendoVentures.DNN.ModuleVue3\bin\UpendoVentures.ModuleVue3.dll
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl  I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/29c2cd4c23f0f71514ed72b29368b850e3788460

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Debug/UpendoVentures.UpendoSendUI.csproj.FileListAbsolute.txt:
```diff
@@ -0,0 +1,15 @@
+D:\Trabajo\NTSPRINT\DNN Modules\Upendo-SendUI\bin\UpendoVentures.UpendoSendUI.dll
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/40b9e90584fdde3a020019d18e8ea336fb28eda3

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Release/Module.csproj.FileListAbsolute.txt:
```diff
@@ -0,0 +1,7 @@
+D:\Trabajo\NTSPRINT\Extension Generator Upendo DNN\Modules\Vue3JS\obj\Release\Module.csproj.CoreCompileInputs.cache
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Release/UpendoVentures.ModuleVue3.csproj.CoreCompileInputs.cache:
```diff
@@ -0,0 +1 @@
+f071406cfc4d1f7783c94cb23474736d0e5436d8
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/6e4e0f9d2b206e3c0bdf8534bf38d807ff90513f

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Release/UpendoVentures.ModuleVue3.csproj.FileListAbsolute.txt:
```diff
@@ -0,0 +1,18 @@
+C:\Users\Nariel\Documents\Upendo\Modules\bin\UpendoVentures.ModuleVue3.dll
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Release/UpendoVentures.Modules.UpendoSendUI.csproj.CoreCompileInputs.cache:
```diff
@@ -0,0 +1 @@
+53962ca4de65c2b0954a08b92f4a0a66732d65ea
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/0e379c6022c8cbd4f955a0485885b57fbbb85e61

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Release/UpendoVentures.Modules.UpendoSendUI.csproj.FileListAbsolute.txt:
```diff
@@ -0,0 +1,7 @@
+D:\Trabajo\NTSPRINT\DNN Modules\Upendo-Send\bin\UpendoVentures.Modules.UpendoSendUI.dll
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/820edd6a9d20407f7975b4966b3fa1102346a3c8

WillStrohl commented on generators/spa/templates/Vue3JS/common/obj/Release/UpendoVentures.UpendoSendUI.csproj.FileListAbsolute.txt:
```diff
@@ -0,0 +1,13 @@
+D:\Trabajo\NTSPRINT\bin\UpendoVentures.UpendoSendUI.dll
```
@alejoroman0605 You ignored this comment.  CC: @ismaelTartola

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/eb733cf3810bc068ec675590ee91ec334db4ee9e

WillStrohl reviewed (changes requested):
@alejoroman0605 Not approved. Many of my comments were simply ignored.  CC: @ismaelTartola

WillStrohl commented on generators/spa/templates/Vue3JS/Module.dnn:
```diff
@@ -0,0 +1,101 @@
+<dotnetnuke type="Package" version="5.0">
+  <packages>
+    <package name="<%= fullNamespace %>" type="Module" version="01.00.00">
```
@alejoroman0605 There is a token for this version number.  https://github.com/UpendoVentures/generator-upendodnn/blob/master/generators/spa/index.js#L121

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/c0437c2115848e91e47ac709f478d2f74ff60337

WillStrohl commented on generators/spa/templates/Vue3JS/Module.dnn:
```diff
@@ -0,0 +1,101 @@
+      <license src="License.txt"></license>
+      <releaseNotes src="ReleaseNotes.txt"></releaseNotes>
+      <dependencies>
+        <dependency type="CoreVersion">09.05.01</dependency>
```
@alejoroman0605 The DNN version number is wrong. It should be 9.8.0.

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/c0437c2115848e91e47ac709f478d2f74ff60337

WillStrohl commented on generators/spa/templates/Vue3JS/Module.dnn:
```diff
@@ -0,0 +1,101 @@
+            <script type="Install">
+              <path>Providers\DataProviders\SqlDataProvider</path>
+              <name>01.00.00.SqlDataProvider</name>
+              <version>01.00.00</version>
```
@alejoroman0605 There is a token for this. https://github.com/UpendoVentures/generator-upendodnn/blob/master/generators/spa/index.js#L121

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/c0437c2115848e91e47ac709f478d2f74ff60337

WillStrohl commented on generators/spa/templates/Vue3JS/Module.dnn:
```diff
@@ -0,0 +1,101 @@
+            <script type="UnInstall">
+              <path>Providers\DataProviders\SqlDataProvider</path>
+              <name>Uninstall.SqlDataProvider</name>
+              <version>01.00.00</version>
```
@alejoroman0605 There is a token for this. https://github.com/UpendoVentures/generator-upendodnn/blob/master/generators/spa/index.js#L121

alejoroman0605 replied:
@WillStrohl I just fixed it https://github.com/UpendoVentures/generator-upendodnn/pull/280/commits/c0437c2115848e91e47ac709f478d2f74ff60337

WillStrohl reviewed (changes requested):
@alejoroman0605 Found a few more issues and reviewed after today's updates.  A lot more work is necessary.

WillStrohl reviewed (approved):
Looks good!  Please remember to create an issue for the `<%= version %>` missing from the Vue 2 template.

WillStrohl commented:
@alejoroman0605  Merging despite a potential false positive on the NewtonSoft.Json dependency being reported as lower than it is.  (security warning)
